### PR TITLE
Introducing preoperation support

### PR DESCRIPTION
**Pull request by MISBMS**
Kindly review and accept the changes.

### CODE REVIEW

File: src/main/java/io/appform/jsonrules/expressions/preoperation/numeric/DivideOperation.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+			throw new IllegalArgumentException("Division by zero not allowed");
+		}
+		return (leftOperand / rightOperand);
+	}
+
+	@Override
+	public double compute(double leftOperand, double rightOperand) {
+		if (Double.isInfinite(leftOperand) || Double.isInfinite(rightOperand)) {
+			throw new IllegalArgumentException("Either of the operands is infinite");
+		} else if (rightOperand == 0) {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/preoperation/numeric/DivideOperation.java]: "At least one of the operands is infinite"

File: src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		case DAY_OF_YEAR:
+			return dateTime.get(ChronoField.DAY_OF_YEAR);
+		case WEEK_OF_MONTH:
+			return dateTime.get(ChronoField.ALIGNED_WEEK_OF_MONTH);
+		case WEEK_OF_YEAR:
+			return dateTime.get(ChronoField.ALIGNED_WEEK_OF_YEAR);
+		case MONTH_OF_YEAR:
+			return dateTime.get(ChronoField.MONTH_OF_YEAR);
+		default:
+			throw new IllegalArgumentException("Operand doesnot represent a valid field");
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java]: Move this to below. Let default fall through.

File: src/main/java/io/appform/jsonrules/expressions/preoperation/date/CalendarOperation.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ */
+@Data
+@EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true)
+public abstract class CalendarOperation extends PreOperation<Object> {
+	private String operand;
+	private String zoneOffSet;
+
+	public CalendarOperation(PreOperationType type) {
+		super(type);
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/preoperation/date/CalendarOperation.java]: Only support protected constructors in base classes.

File: src/main/java/io/appform/jsonrules/expressions/preoperation/date/CalendarOperation.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	
+	public CalendarOperation(PreOperationType type, String operand, String zoneOffSet) {
+		this(type);
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/preoperation/date/CalendarOperation.java]: protected

File: src/main/java/io/appform/jsonrules/expressions/preoperation/date/CalendarOperation.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		this.operand = operand;
+		this.zoneOffSet = zoneOffSet;
+	}
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/preoperation/date/CalendarOperation.java]: Handle null or empty time zone

File: src/main/java/io/appform/jsonrules/expressions/JsonPathBasedExpression.java [hunk 6]
```diff
@@ -33,20 +37,27 @@
...
 
-    protected JsonPathBasedExpression(ExpressionType type, String path) {
+    protected JsonPathBasedExpression(ExpressionType type, String path, PreOperation<?> preoperation) {
         this(type);
         this.path = path;
+        this.preoperation = preoperation;
     }
-
+    
     @Override
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/JsonPathBasedExpression.java]: Remove the empty line

File: src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+package io.appform.jsonrules.utils;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoField;
+
+public class PreOperationUtils {
+
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java]: Make this an interface or add a private constructor.

File: src/test/java/io/appform/jsonrules/PreOperationTest.java [hunk 8]
```diff
@@ -0,0 +1,1368 @@
...
+	            .build()
+	            .evaluate(context);
+        	Assert.fail("Should have thrown an exception");
+        } catch(IllegalArgumentException e) {
+        	Assert.assertTrue("Doesn't represent a valid date time string", true);
+        }
+        
+
+		// EpochOperation
+        Assert.assertTrue(EqualsExpression.builder()
```
Comment (santanusinha) [inline src/test/java/io/appform/jsonrules/PreOperationTest.java]: Create separate tests for different op types

File: src/test/java/io/appform/jsonrules/PreOperationTest.java [hunk 9]
```diff
@@ -0,0 +1,1368 @@
...
+                .evaluate(context));
+        
+        Assert.assertFalse(EqualsExpression.builder()
+        		.path("/dateTime")
+        		.preoperation(EpochOperation.builder().operand("minute_of_hour").build())
+        		.value(dateTime.atOffset(ZoneOffset.UTC).get(ChronoField.MINUTE_OF_HOUR))
+        		.build()
+        		.evaluate(context));
+	}
+
```
Comment (santanusinha) [inline src/test/java/io/appform/jsonrules/PreOperationTest.java]: Add tests that use the serialized version of the ops to check serialization/deserialization issues.

File: src/test/java/io/appform/jsonrules/RuleTest.java [hunk 10]
```diff
@@ -69,6 +69,6 @@ public void testRepresentation() throws Exception {
         final String ruleRep = rule.representation(mapper);
 
         System.out.println(ruleRep);
-        Assert.assertEquals("{\"type\":\"not\",\"children\":[{\"type\":\"or\",\"children\":[{\"type\":\"less_than\",\"path\":\"/value\",\"value\":11},{\"type\":\"greater_than\",\"path\":\"/value\",\"value\":30}]}]}", ruleRep);
+        Assert.assertEquals("{\"type\":\"not\",\"children\":[{\"type\":\"or\",\"children\":[{\"type\":\"less_than\",\"path\":\"/value\",\"preoperation\":null,\"value\":11},{\"type\":\"greater_than\",\"path\":\"/value\",\"preoperation\":null,\"value\":30}]}]}", ruleRep);
     }
```
Comment (santanusinha) [inline src/test/java/io/appform/jsonrules/RuleTest.java]: This change should not be required.

File: src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java (hunk 7 above)
Comment (MISBMS) [inline src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java]: Done

**Comment by MISBMS** [comment]
Review changes incorporated. Please check and merge. Thanks.

**Pull request merged**